libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -666,56 +666,134 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+	// Fetch inherited functions and sort them by signature.
+	// We get at least one function per signature and direct base contract, which is
+	// enough because we re-construct the inheritance graph later.
+	FunctionMultiSet nonOverriddenFunctions = inheritedFunctions(_contract);
+	// Remove all functions that match the signature of a function in the current contract.
+	nonOverriddenFunctions -= _contract.definedFunctions();
+
+	// Walk through the set of functions signature by signature.
+	for (auto it = nonOverriddenFunctions.cbegin(); it != nonOverriddenFunctions.cend();)
 	{
-		auto [begin,end] = inheritedFuncs.equal_range(*it);
+		static constexpr auto compareById = [](auto const* a, auto const* b) { return a->id() < b->id(); };
+		std::set<FunctionDefinition const*, decltype(compareById)> baseFunctions(compareById);
+		for (auto nextSignature = nonOverriddenFunctions.upper_bound(*it); it != nextSignature; ++it)
+			baseFunctions.insert(*it);
 
-		// Only one function
-		if (next(begin) == end)
+		if (baseFunctions.size() <= 1)
 			continue;
 
-		// Not an overridable function
-		if ((*it)->isConstructor())
+		// Construct the override graph for this signature.
+		// Reserve node 0 for the current contract and node
+		// 1 for an artificial top node to which all override paths
+		// connect at the end.
+		struct OverrideGraph
 		{
-			for (begin++; begin != end; begin++)
-				solAssert((*begin)->isConstructor(), "All functions in range expected to be constructors!");
-			continue;
-		}
+			OverrideGraph(decltype(baseFunctions) const& _baseFunctions)
+			{
+				for (auto const* baseFunction: _baseFunctions)
+					addEdge(0, visit(baseFunction));
+			}
+			std::map<FunctionDefinition const*, int> nodes;
+			std::map<int, FunctionDefinition const*> nodeInv;
+			std::map<int, std::set<int>> edges;
+			int numNodes = 2;
+			void addEdge(int _a, int _b)
+			{
+				edges[_a].insert(_b);
+				edges[_b].insert(_a);
+			}
+		private:
+			/// Completes the graph starting from @a _function and
+			/// @returns the node ID.
+			int visit(FunctionDefinition const* _function)
+			{
+				auto it = nodes.find(_function);
+				if (it != nodes.end())
+					return it->second;
+				int currentNode = numNodes++;
+				nodes[_function] = currentNode;
+				nodeInv[currentNode] = _function;
+				if (_function->overrides())
+					for (auto const* baseFunction: _function->annotation().baseFunctions)
+						addEdge(currentNode, visit(baseFunction));
+				else
+					addEdge(currentNode, 1);
+
+				return currentNode;
+			}
+		} overrideGraph(baseFunctions);
 
-		// Function has been explicitly overridden
-		if (contains_if(
-			contractFuncs,
-			[&] (FunctionDefinition const* _f) {
-				return hasEqualNameAndParameters(*_f, **it);
+		// Detect cut vertices following https://en.wikipedia.org/wiki/Biconnected_component#Pseudocode
+		// Can ignore the root node, since it is never a cut vertex in our case.
```

Comment:
Why is it never a cut vertex? Since the graph is undirected, if the root function overrides two functions that are in different subgraphs it will be a cut vertex too.

Reply:
It's maybe better to say that we ignore the root node because we are interested in "base" cut vertices, or something like that.

Reply:
Since we have the artificial vertex every path connects to in the end, all children of the root will always be in the same subgraph. So the only case in which the root is a cut vertex is the case that it overrides only one function, but that's excluded in the outermost if.

Reply:
Ok!